### PR TITLE
Add new transformers model type: Bart

### PR DESCRIPTION
**Pull request by wangyems**
**Description**: Describe your changes.
issues that apply:
https://github.com/microsoft/onnxruntime/issues/8637
https://github.com/microsoft/onnxruntime/issues/7796
-enable the a new type as "bart" in transformer optimizer
-enhance the reshape fusion for huggingface(4.9) bart model
-fuse the self-attention in encoder

Linux/V-100/intra_num_threads=24

**Basically benefit cpu with smaller size. For large size one could disable attention fusion**

**Before**:
| model | EP | precision | b1_s128 | b1_s512 | b2_s128 | b2_s512 |
| ------- | -------  | -------  | -------  | -------  | -------  | -------  | 
| bart-base | cpu | fp32 | 71.94 | 339.22 | 131.94 | 432.31 |
| bart-large | cpu | fp32 | 213.95 | 656.69 | 377.22 | 1287.50 |
| bart-base | cuda | fp32 | 5.30 | 14.67 |10.89 | 25.53 | 
| bart-large | cuda | fp32 |13.82 |43.64 |21.25 | 81.58 |

**After**:
| model | EP | precision | b1_s128 | b1_s512 | b2_s128 | b2_s512 |
| ------- | -------  | -------  | -------  | -------  | -------  | -------  | 
| bart-base | cpu | fp32 | 59.73 | 219.49 | 102.28 | 414.53 |
| bart-large | cpu | fp32 | 189.58 | 699.35 | 328.35 | 1323.91 |
| bart-base | cuda | fp32 | 5.12 | 14.61 |10.00 | 25.43 | 
| bart-large | cuda | fp32 |13.22 | 43.29 | 20.89 | 81.12 |

TODO: (Will include in the separate PR when get to them)
-accepts attention masks
-implement cross-attention in decoder
-fuse the decoder attention

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/python/tools/transformers/fusion_reshape.py [hunk 1]
```diff
@@ -31,16 +48,43 @@ def fuse(self, reshape_node, input_name_to_nodes, output_name_to_node):
...
 
+        shape = []
+        gather_value = self.model.get_constant_value(gather_0.input[1])
+        if gather_value == 0:
+            shape.append(0)
+
         path1 = self.model.match_parent_path(concat_node, ['Unsqueeze', 'Gather', 'Shape'], [1, 0, 0],
                                              output_name_to_node)
         if path1 is None:
+            # Adjust for Bart
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/fusion_reshape.py:59]: These reshape fusion is not generic. I think it is better to put to preprocess of onnx_model_bart.py.

File: onnxruntime/python/tools/transformers/fusion_reshape.py [hunk 2]
```diff
@@ -31,16 +48,43 @@ def fuse(self, reshape_node, input_name_to_nodes, output_name_to_node):
...
+            if len(concat_node.input) == 4:
+                input_1_proto = self.model.get_initializer(concat_node.input[1])
+                input_2_proto = self.model.get_initializer(concat_node.input[2])
+                input_3_proto = self.model.get_initializer(concat_node.input[3])
+                if input_1_proto is None or input_2_proto is None or input_3_proto is None:
+                    return
+                input_1 = numpy_helper.to_array(input_1_proto)
+                input_2 = numpy_helper.to_array(input_2_proto)
+                input_3 = numpy_helper.to_array(input_3_proto)
+                if input_1[0] == -1 and input_2[0] > 0 and input_3[0] > 0:
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/fusion_reshape.py:69]: also check len(input_1) == 1 and len(input_2) == 1 and len(input_3) == 1?

**Comment by yufenglee** [comment]
Do you perf improvement for fp16 on GPU? And why there is no improvement for fp32 on GPU?

File: onnxruntime/python/tools/transformers/fusion_reshape.py (hunk 1 above)
Comment (wangyems) [reply onnxruntime/python/tools/transformers/fusion_reshape.py:59]: This reshape fusion is based on the assumption that hidden_state is >2d and gemm won't change the first and second dimension. So I think it's generic to transformer model?

File: onnxruntime/python/tools/transformers/fusion_reshape.py (hunk 2 above)
Comment (wangyems) [reply onnxruntime/python/tools/transformers/fusion_reshape.py:69]: good catch!

**Comment by wangyems** [comment]
> Do you perf improvement for fp16 on GPU? And why there is no improvement for fp32 on GPU?

Seems the GPU perf does not benefit a lot/always benefit from attention fusion. My observation is that in optimized model, the SequentialExecutor takes less time but the ratio of time on model_run over time on sequential_executor is bigger which offset part of result from optimization

**Review by tianleiwu — APPROVED** [review]

**Pull request merged**